File: src/ROSE_callSuper.R
#!/usr/bin/env Rscript

suppressMessages(library(data.table))
suppressMessages(library(docstring))
suppressMessages(library(optparse))


get_path <- function() {
	#' Get path to ROSE tool directory
	#' 
	#' @return Path to ROSE tool directory

	path <- strsplit(getwd(), "/")
	return(paste(unlist(path)[1:which(unlist(path)=="ROSE")], collapse="/"))
}


option_list <- list(
	make_option(c("-o", "--output"), action="store", type="character", default=NULL, help="Output directory name"),
	make_option(c("-d", "--density"), action="store", type="character", default=NULL, help="Stitched enhancer loci signal density file"),
	make_option(c("-g", "--gff"), action="store", type="character", default=NULL, help="Constituent enhancers peak file"),
	make_option(c("-c", "--control"), action="store", type="character", default=NULL, help="Control .bam file")
)

#Parse arguments from command line
opt <- parse_args(OptionParser(option_list=option_list))

#Initialising variables
path <- get_path()
densityName <- tail(unlist(strsplit(opt$density, "/")))

#Load necessary functions
source(paste(c(path, "src/superenhancers/output.R"), collapse="/"))
source(paste(c(path, "src/superenhancers/super_enhancer.R"), collapse="/"))

#Read stitched enhancer loci density signal file as dataframe
stitched_regions <- fread(opt$density, sep="\t")

#Subtract background signal if control is available
if (is.null(opt$control)) {
	rankBy_vector = stitched_regions[,7]
} else {
	rankBy_vector = stitched_regions[,7] - stitched_regions[,8]
}

#Setting negative values to 0
rankBy_vector[rankBy_vector < 0] <- 0

#Calculating the superenhancer signal cut-off value
cutoff_options <- calculate_cutoff(as.matrix(rankBy_vector))

#Get superenhancers based on cut-off
superEnhancerRows <- which(rankBy_vector > cutoff_options$absolute)
typicalEnhancers <- setdiff(1:nrow(stitched_regions), superEnhancerRows)
enhancerDescription <- paste(basename(opt$gff)," Enhancers\nCreated from ", tail(unlist(strsplit(opt$density, "/")), n=1), "\nRanked by ", colnames(stitched_regions)[7], "\nUsing cutoff of ", cutoff_options$absolute, " for Super-Enhancers", sep="", collapse="")

#Get base .gff file name
gff <- tools::file_path_sans_ext(basename(opt$gff))

#Creating hockey stick plot
hockey_stick_plot(rankBy_vector, cutoff_options, superEnhancerRows, opt$output, gff, opt$control, colnames(stitched_regions)[7])

#Rank stitched enhancer loci by background signal corrected density signal and output to .bed file
bedFileName = paste(opt$output, "/", gff, "_Enhancers_withSuper.bed", sep="")
convert_stitched_to_bed(stitched_regions, paste(colnames(stitched_regions)[7], "Enhancers"), enhancerDescription, bedFileName, densitySignal=as.matrix(rankBy_vector), splitSuper=TRUE, superRows=superEnhancerRows)

#Output stitched enhancer loci ranking
convert_stitched_to_gateway_bed(stitched_regions, paste(opt$output, "/", gff, sep=""), splitSuper=TRUE, densitySignal=as.matrix(rankBy_vector), superRows=superEnhancerRows)

#Select super enhancers
true_super_enhancers <- stitched_regions[superEnhancerRows,]

#Create matrix of stitched enhancer loci rankings and super status
data = c(nrow(stitched_regions)-rank(rankBy_vector, ties.method="first")+1, ifelse(1:nrow(stitched_regions) %in% superEnhancerRows, 1, 0))
additionalTableData <- matrix(data=data, ncol=2, nrow=nrow(stitched_regions))
colnames(additionalTableData) <- c("enhancerRank", "isSuper")

#Output rankings and status matrix
enhancerTableFile = paste(opt$output, "/", gff, "_AllEnhancers.table.txt", sep="")
superTableFile = paste(opt$output, "/", gff, "_SuperEnhancers.table.txt", sep="")

writeSuperEnhancer_table(stitched_regions, enhancerDescription, enhancerTableFile, additionalData=additionalTableData)
writeSuperEnhancer_table(true_super_enhancers, enhancerDescription, superTableFile, additionalData=additionalTableData[superEnhancerRows,])








	

File: src/ROSE_callSuper.py
#!/usr/bin/env python

import numpy as np
import pandas as pd

from pathlib import Path
from scipy.stats import rankdata
from src.superenhancers.output import (
    convert_stitched_to_bed,
    hockey_stick_plot,
    write_enhancer_table
)
from src.superenhancers.super_enhancer import calculate_cutoff
from src.utils.file_helper import check_file, check_path
from typing import Union


def get_super(
    output: str,
    density: str,
    gff: str,
    control: Union[str, None],
) -> None:
    """Find superenhancers among stitched enhancer loci
    """

    # Read stitched enhancer loci density signal file as dataframe
    stitched_regions = pd.read_csv(check_file(density), sep="\t")
    rank_cols = list(stitched_regions.columns[:6])


    # Subtract control signal if control is available
    if control:
        rankBy = stitched_regions.iloc[:, 6:-1].sub(
            stitched_regions.iloc[:, -1],
            axis=0
        )
    else:
        rankBy = stitched_regions.iloc[:, 6:]


    # Loop over each bam file density
    for bam in rankBy.columns:

        # Dropping other bam file densities from the dataframe
        cols = rank_cols + [bam] + [stitched_regions.columns[-1]]
        stitched_regions_temp = stitched_regions.filter(cols)

        # Setting negative values to 0
        rankBy_vector = rankBy[bam].copy()
        rankBy_vector[rankBy_vector < 0] = 0

        # Calculate the superenhancer density signal cut-off value
        y_cutoff = calculate_cutoff(np.asarray(rankBy_vector.copy()))

        # Get superenhancers based on their density signal
        superEnhancerRows = np.where(rankBy_vector > y_cutoff)[0]

        # Creating hockey stick plot
        hockey_stick_plot(
            np.asarray(rankBy_vector).copy(),
            y_cutoff,
            superEnhancerRows,
            output,
            gff,
            control,
            bam
        )

        # Create output file header
        enhancerDescription = f"{Path(gff).name} Enhancers\nCreated from {Path(density).name}"
        enhancerDescription += f"\nRanked by {bam}\nUsing cutoff of {y_cutoff} for Super-Enhancers"

        # Rank stitched enhancer loci by control corrected density signal and
        # output to .bed file
        bedName = str(Path(output, f"{Path(gff).stem}_{Path(bam).stem}_enhancers_withSuper.bed"))
        convert_stitched_to_bed(
            stitched_regions_temp,
            enhancerDescription,
            bedName,
            np.asarray(rankBy_vector),
            superEnhancerRows,
            bam
        )

        # Calculate stitched enhancer loci rankings and super status
        enhancer_rank = len(stitched_regions)-rankdata(rankBy_vector, method="ordinal")+1
        super_status = [
            1 if sr in superEnhancerRows else 0
            for sr in range(0, len(stitched_regions))
        ]
        additional_data = pd.DataFrame(
            {"enhancerRank": enhancer_rank, "isSuper": super_status}
        )

        # Output rankings and status dataframe
        enhancer_file = check_path(
            Path(output, f"{Path(gff).stem}_{Path(bam).stem}_AllEnhancers.table.txt")
        )
        write_enhancer_table(
            stitched_regions_temp,
            enhancerDescription,
            enhancer_file,
            additional_data
        )

        super_file = check_path(
            Path(output, f"{Path(gff).stem}_{Path(bam).stem}_SuperEnhancers.table.txt")
        )
        write_enhancer_table(
            stitched_regions_temp.iloc[superEnhancerRows, :],
            enhancerDescription,
            super_file,
            additional_data.iloc[superEnhancerRows, :]
        )
